File: app/src/main/java/com/qalliance/treetracker/TreeTracker/database/DatabaseManager.java
package com.qalliance.treetracker.TreeTracker.database;

import android.content.ContentValues;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;

/**
 * Created by lei on 11/9/17.
 */

public class DatabaseManager {

    private static DatabaseManager sInstance;
    private DbHelper mDbHelper;
    private SQLiteDatabase mDatabase;
    private int mOpenCounter;

    public static synchronized DatabaseManager getInstance(DbHelper dbHelper) {
        if (sInstance == null) {
            sInstance = new DatabaseManager(dbHelper);
        }

        return sInstance;
    }

    private DatabaseManager(DbHelper dbHelper) {
        mDbHelper = dbHelper;
    }

    public synchronized SQLiteDatabase openDatabase() {
        mOpenCounter++;
        if(mOpenCounter == 1) {
            // Opening new database
            mDatabase = mDbHelper.getWritableDatabase();
        }
        return mDatabase;
    }

    public synchronized void closeDatabase() {
        mOpenCounter--;
        if(mOpenCounter == 0) {
            // Closing database
            mDatabase.close();
        }
    }

    public Cursor queryCursor(String sql, String[] selectionArgs) {
        return mDatabase.rawQuery(sql, selectionArgs);
    }

    public long insert(String table, String nullColumnHack, ContentValues contentValues) {
        return mDatabase.insert(table, nullColumnHack, contentValues);
    }

    public int update(String table, ContentValues values, String whereClause, String[] whereArgs) {
        return mDatabase.update(table, values, whereClause, whereArgs);
    }

    public int delete(String table, String whereClause, String[] whereArgs) {
        return mDatabase.delete(table, whereClause, whereArgs);
    }
}

File: app/src/main/java/org/greenstand/android/TreeTracker/database/DatabaseManager.kt
package org.greenstand.android.TreeTracker.database

import android.content.ContentValues
import android.database.Cursor
import android.database.sqlite.SQLiteDatabase
import androidx.sqlite.db.SupportSQLiteDatabase
import androidx.sqlite.db.SupportSQLiteOpenHelper

/**
 * Created by lei on 11/9/17.
 */

class DatabaseManager(private val dbHelper: SupportSQLiteOpenHelper) {

    private var database: SupportSQLiteDatabase? = null

    @Synchronized
    fun openDatabase(): SupportSQLiteDatabase {
        if (database == null) {
            // Opening new database
            database = dbHelper.writableDatabase
        }
        return database!!
    }


    fun queryCursor(sql: String, selectionArgs: Array<String>?): Cursor {
        return database!!.query(sql, selectionArgs)
    }

    fun insert(table: String, nullColumnHack: String?, contentValues: ContentValues): Long {
        return database!!.insert(table, SQLiteDatabase.CONFLICT_REPLACE, contentValues)
    }

    fun update(table: String, values: ContentValues, whereClause: String, whereArgs: Array<String>): Int {
        return database!!.update(table, SQLiteDatabase.CONFLICT_REPLACE, values, whereClause, whereArgs)
    }

    fun delete(table: String, whereClause: String, whereArgs: Array<String>): Int {
        return database!!.delete(table, whereClause, whereArgs)
    }


}
